Add target configs for cross-compiling to Windows using Clang-CL and Microsoft CRT

erikbs opened the pull request:
This adds configurations for cross-compiling to Windows (x86_64 or aarch64) using Clang-CL and other LLVM tools with MSVC-like syntax (llvm-link, llvm-rc, llvm-lib). They use the Microsoft CRT and Windows SDK instead of the mingw toolchain and libraries. A few adjustments to the Unix Makefile template had to be made for this to work, controlled by an option `msvc_syntax`. I tried to make the changes as little intrusive as possible.

I found no obvious pattern in the configuration names. Please let me know if I should name them differently.

Using the existing procedures for generating GCC-/Clang-compatible Assembly, these configurations do not require `no-asm`. Also, `clang-cl` emits warnings about unused parameters when processing assembly files, but I considered this acceptable since doing something with it would probably require more conditional changes. The goal has been to generate output as close to the VC-WIN* configurations as possible.

I have tested this primarily on an M4 Mac, cross compiling mainly from aarch64 to x86_64, with additional testing in Docker on a GNU/Linux system. The Microsoft CRT and Windows SDK were obtained using xwin.

To try to make sure that I did not break anything, I have also tested configurations `mingw64` (MSYS2/Windows), `VC-WIN64A` (Windows with MSVC 2022), `darwin64-x86_64` (macOS 15.5) and `linux-x86_64` (RHEL 9.6).

Note: the `mingw64` configuration did not build successfully because of some very recent changes to `bioprinttest.c`, but that is not related to my changes. I verified that by trying to build from master. If commenting out the lines in that file referencing unknown symbols, the build completes successfully both with and without my changes.

The purpose of these configurations is to enable cross-compilation from macOS or GNU/Linux to Windows without having to use the mingw-based configurations, but I guess it would be possible to use them from e.g. MSYS2 on Windows as well. I had two main reasons for not building on Windows using the VC-WIN* configurations: build times (building on Windows is painfully slow, even with Ccache) and the possibility to build for multiple targets on a single host.

Comparing the libraries and binaries compiled natively against the cross-compiled ones, I observed that the cross-compiled ones generally were a bit larger, except for `libcrypto*.dll`, which was significantly smaller, but none of those deviations were unreasonable. When comparing the outputs of `openssl.exe speed`, the cross-compiled binary fared worse on smaller block sizes in the first tests (the native version computed up to 40-50 % more hashes in 3 seconds), but then in the following tests the cross-compiled version quite consistently outperformed the native version by a large margin (for example hmac and AES), often reaching 2-2,5 times more operations than the native version. I ran both binaries on the same Windows system, one at the time.

This is my first attempt at diving into the OpenSSL build system and I have fairly limited experience with Perl, so there may well be room for improvement. A few questions:
* I added my target configuration file as `15-wincross.conf`, using `15` because I likened it to the Android and iOS configs. Should I have used another number instead?
* Should I have set the `multilib` option? It caused the libraries to be put in `usr/lib-x64`, which looked non-standard to me, so I skipped that option.
* To avoid conflict between static `.lib` files and shared import `.lib` files, I set the extension of static libraries to `_static.lib`, inspired by the Windows scripts. Unfortunately this adds `_static` to a couple of more libraries than when building on Windows, but I am not sure if it matters. There is a comment in Unix.pm saying that this `_static` prefix might be added in the future, but I thought that doing that change now would affect too many other configurations and carry too much risk, so I went with putting the suffix in my build configurations instead. Is this an acceptable solution?
* I added an `msvc_syntax` option that alters the behaviour of certain things in the Unix Makefile template. Is this acceptable or too intrusive in the Unix Makefile template? And is it okay to have one option control three different things (parameter order of the linker command, workaround for lib.exe/llvm-lib not supporting appending objects to archive and special treatment of .rc files because rc.exe/llvm-rc is much less flexible than GNU windres) or should I have made three different options instead?

---

Given that Docker is installed, my suggested solution can be tested by running this script in an empty folder on a Linux x86_64 host:

```bash
#!/usr/bin/env bash

# Download the Microsoft CRT and Windows SDK
docker run --rm -it -u $(id -u):$(id -g) -v "$(pwd):$(pwd)" -w "$(pwd)" axemsolutions/xwin:0.6.2 \
    xwin --accept-license --arch x86_64 splat --preserve-ms-arch-notation --include-debug-libs --output x86_64

# Clone my branch
git clone --depth 1 -b feature/win64-cross-clang-cl https://github.com/erikbs/openssl

# Set environment variables, configure and build
xwindir=$(pwd)/x86_64
CFLAGS="-vctoolsdir ${xwindir}/crt -winsdkdir ${xwindir}/sdk"
LDFLAGS="/vctoolsdir:${xwindir}/crt /winsdkdir:${xwindir}/sdk"
docker run --rm -it -u $(id -u):$(id -g) -v "$(pwd):$(pwd)" -w "$(pwd)/openssl" \
    -e CFLAGS="$CFLAGS" -e LDFLAGS="$LDFLAGS" silkeh/clang:19 \
    bash -c "./Configure win64-x86_64-cross-clang-cl && make -j32"
```

erikbs committed:
mkdef.pl: add alias win => WINDOWS

erikbs committed:
x86_64-xlate.pl: add flavour "win64", same as "mingw64" except that it does not set $prefix

erikbs committed:
unix-Makefile.tmpl: add two-step recipe for .rc files (preprocess with CC, compile with RC)

erikbs committed:
unix-Makefile.tmpl: use rsp file instead of multiple calls to AR for MSVC-like tools

erikbs committed:
unix-Makefile.tmpl: add support for MSVC-like linker syntax (link.exe, llvm-link)

erikbs committed:
build.info: treat configurations win64-* like VC-*

erikbs committed:
Add new configuration file 15-wincross.conf for cross-compilation to Windows/MSVC using clang-cl

erikbs committed:
unix-Makefile.tmpl: add "win64" configurations to windowsdll

t8m commented:
Would you be willing to become a community maintainer of these newly added platforms? We do not want to add new build targets without having a community maintainer associated with them.

t8m commented:
> @t8m I can consider that, but what does it imply? If it’s building OpenSSL every once in a while and submitting patches whenever something breaks, it should be manageable. If it also involves providing user support or actively monitoring the commit history on `master` to look for things that might break these platform configs, it’s probably more than I’m comfortable handling.

It's rather the first thing although if users report issues on these platforms we will also ping you on them.

erikbs committed:
Add new Perl platform Windows::Unix

erikbs committed:
Assembly generation scripts: set $win64=1 for flavour "win64"

erikbs committed:
unix-Makefile.tmpl: clean .pdb and .lib files for win64 targets

erikbs commented:
Some final adjustments pushed (new rebase on master, fix typos in commit messages, treat "win64" as "mingw64" in assembly generation scripts to fix a segfault encountered when using libcrypto in an application).

solomoncyj commented:
Is it possible to have the win x64 one to have a msvc-clang mode?

solomoncyj commented:
https://github.com/llvm/llvm-project/releases/tag/llvmorg-21.1.2 i mean like, using the llvm ecosystem in windows

erikbs committed:
15-wincross.conf: enable uplink

erikbs committed:
unix-Makefile.tmpl: make max_per_call overridable and override to 250 in 15-wincross.conf

erikbs committed:
unix-Makefile.tmpl: use the rsp file solution from obj2lib in obj2shlib as well

erikbs commented:
I pushed three more commits, one to enable uplink again and two to fix the linker problems in MSYS2. With this, it should be possible to use these configurations in MSYS2 on Windows with Visual Studio installed (so perhaps not exactly cross-compilation).

Also, I tested adding these two more configurations to `Configuration/50-win-clang-cl.conf`:
```
    "VC-WIN64-CLANGASM" => {
        inherit_from    => [ "VC-WIN64A" ],
        AS              => "clang-cl.exe",
        ASFLAGS         => "/nologo /Zi --target=x86_64-pc-windows-msvc",
        asflags         => "/c",
        asoutflag       => "/Fo",
        perlasm_scheme  => "win64",
    },
    "VC-CLANG-WIN64-CLANGASM" => {
        inherit_from    => [ "VC-WIN64-CLANGASM" ],
        CC              => "clang-cl.exe",
    }
```
and it seems that both work. I did not commit this to the same PR, but it indicates that it should be possible to use Clang for everything when targeting x86_64 too.

erikbs commented:
The possibility of using Ccache is definitely a great advantage that I actually had in mind, since I’m building OpenSSL quite often as part of a software suite that is rebuilt each time one of the components is updated. I know that it takes roughly half an hour to build OpenSSL on that system today (using `VC-WIN64A`), whereas the Linux version takes significantly less time to build. `nmake` is of course a big part of that picture, but I assume that speedup can be achieved with Ccache as well. Out of curiosity I decided to do some measurements. I don’t have access to the build server that spends half an hour on OpenSSL right now, so I did the measurements on the computers I used for working on this PR.

Ccache didn’t do much on the M4/aarch64 Mac I used most of the time, though, but then I used it because it has a quite powerful CPU, so I didn’t have to wait forever when tweaking the .conf file:
```bash
$ ./Configure win64-x86_64-cross-clang-cl &>/dev/null && time make -j32 &>/dev/null
real    0m27,731s
user    5m38,486s
sys     1m17,242s

$ git clean -fdx
$ export CC="ccache $CC"
$ export AS="ccache $AS"
$ ./Configure win64-x86_64-cross-clang-cl &>/dev/null && time make -j32 &>/dev/null
real    0m34,726s
user    5m39,972s
sys     1m24,734s

$ git clean -fdx
$ ./Configure win64-x86_64-cross-clang-cl &>/dev/null && time make -j32 &>/dev/null
real    0m34,996s
user    5m44,156s
sys     1m23,597s
```

There were, however, significant improvements when running in MSYS2 on the older, more mid-range x86_64-based Windows computer:
```bash
$ ./Configure win64-x86_64-cross-clang-cl &>/dev/null && time make -j32 &>/dev/null
real    3m6.969s
user    2m19.812s
sys     2m16.124s

$ git clean -fdx
$ export CC="ccache $CC"
$ export AS="ccache $AS"
$ ./Configure win64-x86_64-cross-clang-cl &>/dev/null && time make -j32 &>/dev/null
real    3m35.694s
user    2m16.671s
sys     2m14.489s

$ git clean -fdx
$ ./Configure win64-x86_64-cross-clang-cl &>/dev/null && time make -j32 &>/dev/null
real    1m23.615s
user    2m9.466s
sys     2m2.144s
```
So that’s 2,2x faster with Ccache with 100 % hit rate compared to no Ccache at all. Running in the Developer Command Prompt for VS 2022 on the same computer gives an even greater speedup (calculated duration comments inserted for convenience):
```bat
> type ..\exec.bat
@echo off
git clean -fdx
PERL Configure VC-CLANG-WIN64-CLANGASM >NUL 2>&1
ECHO %TIME%
NMAKE >NUL 2>&1
ECHO %TIME%

> ..\exec.bat
22:14:18,32
22:35:51,80
:: Duration: 00:21:33,48

> set CC=ccache" "clang-cl.exe
> set AS=ccache" "clang-cl.exe
> ..\exec.bat
23:08:39,72
23:37:21,37
:: Duration: 00:28:41,65

> ..\exec.bat
23:40:18,82
23:46:20,64
:: Duration: 00:06:01,82
```
Here, running with Ccache with 100 % hit rate is 3,6x times faster than not using Ccache at all. Corresponding build time with MSVC+NASM (`VC-WIN64A`) without Ccache was `00:18:03,42`.

I didn’t measure the speedup for assembly files alone, but I guess that some time can be saved by using Ccache.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

erikbs commented:
Rebased on master, verified that compilation still works (I tested macOS aarch64 -> Windows x86_64 only). Ready for review.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

andrewkdinh commented:
Please rebase

erikbs commented:
Rebased on master, with some adaptations.

In June, PR #23077 was merged, adding use of a response file to `unix-Makefile.tmpl`. I too did that in my proposal, but I did it differently. The solution on master uses `$(file …)`, but it was noted in https://github.com/openssl/openssl/pull/23077#issuecomment-1929192402 that that is “a GNU make thing” (and I am indeed having problems with it on macOS, but I can work around the problem using `gmake` from MacPorts instead of the ancient `make` binary shipped with the OS). In my original proposal I used a pure Perl solution. I discarded that as part of the rebase and kept the solution on master instead (to reduce the diff and because my solution depended on generating the response file by adding object file names in chunks, so the `$file( ...)` solution is more elegant). I was tempted to change the .lst extension to .rsp (used by MSVC), but I am unsure if that is fine for GCC-like toolchains, so I kept .lst and it seems to work fine with Clang-CL.

The same PR introduced a function `run_on_windows` and used that as condition for when to use a response file. I renamed this function to `use_response_file` and expanded it to include MSVC-like toolchains regardless of host OS. This allows me to use a response file with Clang-CL, but with slightly different ordering of the compiler arguments than with compilers using GCC-like syntax.

~~Then I also used this function as condition for using a response file with `AR`, instead of using my `msvc_syntax` variable directly, as I originally did. The consequence of this is that when running in a Unix-like environment on Windows (Cygwin/Msys), a response file will be used with `AR`. From what I read, both GNU ar and LLVM ar support response files, so I suppose this change is safe for most targets. If, however, there is a toolchain that can be run on Windows in a Unix-like environment that does _not_ support response files, support for it will be broken by this change. If desired, I can revert to using a response file only when `msvc_syntax` equals 1 (completely safe since that is a variable introduced in this PR).~~ I reverted this part. I overlooked (and did not remember) that I had to adjust the syntax as well for it to work with MSVC, so this code would of course not be compatible with Cygwin/Mingw.

I have tested the rebased solution on macOS, which is what I have available at the moment. I can test it again on Linux and in Git Bash (msys) on Windows too, but that will have to wait a day or two.

andrewkdinh commented on test/build.info:
This also needs to be updated:

https://github.com/openssl/openssl/blob/e7c84542e25bca93ec2d3e4819433b8b5e0c483f/test/build.info#L603

```c
&& $config{target} !~ /^(?:VC-|mingw|BC-|Cygwin|win64-)/i
```

erikbs replied:
Thanks! I have amended that commit and then searched all `build.info` files for `VC-` and `mingw` to see if there were more cases added after the last time I rebased on master (there were not as far as I can see).

andrewkdinh commented on Configurations/15-wincross.conf:
```diff
@@ -0,0 +1,59 @@
+
+        bin_lflags       => '/link',
+
+        max_objects_per_link_call => 250,
```
Was this just randomly chosen?

andrewkdinh commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -1812,10 +1828,16 @@ EOF
               my $l = basename($_);
-              $l =~ s/^lib//;
-              $l = "-l" . $l;
+              if ($target{msvc_syntax} // 0) {
```
Does this need to be added for `obj2dso` as well?
